libkbfs/folder_block_ops.go

```diff
@@ -1089,67 +1155,53 @@ func (fbo *folderBlockOps) writeDataLocked(
 		}
 
 		oldLen := len(block.Contents)
-		nCopied += bsplit.CopyUntilSplit(block, !more, data[nCopied:],
-			off+nCopied-startOff)
-
-		// the block splitter could only have copied to the end of the
-		// existing block (or appended to the end of the final block), so
-		// we shouldn't ever hit this case:
-		if more && oldLen < len(block.Contents) {
-			return WriteRange{}, nil, BadSplitError{}
+		max := len(data)
+		if nextBlockOff > 0 {
+			if room := int(nextBlockOff - (off + nCopied - startOff)); room < max {
+				max = room
+			}
 		}
+		nCopied += bsplit.CopyUntilSplit(block, nextBlockOff < 0, data[nCopied:max],
+			off+nCopied-startOff)
 
 		// TODO: support multiple levels of indirection.  Right now the
 		// code only does one but it should be straightforward to
 		// generalize, just annoying
 
 		// if we need another block but there are no more, then make one
-		if nCopied < n && !more {
+		if nCopied < n && nextBlockOff < 0 {
 			// If the block doesn't already have a parent block, make one.
 			if ptr == file.tailPointer() {
-				// pick a new id for this block, and use this block's ID for
-				// the parent
-				newID, err := fbo.config.Crypto().MakeTemporaryBlockID()
+				fblock, err = fbo.createInderectBlockLocked(md, file, uid,
+					DefaultNewBlockDataVersion(fbo.config, false))
 				if err != nil {
 					return WriteRange{}, nil, err
 				}
-				fblock = &FileBlock{
-					CommonBlock: CommonBlock{
-						IsInd: true,
-					},
-					IPtrs: []IndirectFilePtr{
-						{
-							BlockInfo: BlockInfo{
-								BlockPointer: BlockPointer{
-									ID:       newID,
-									KeyGen:   md.LatestKeyGeneration(),
-									DataVer:  fbo.config.DataVersion(),
-									Creator:  uid,
-									RefNonce: zeroBlockRefNonce,
-								},
-								EncodedSize: 0,
-							},
-							Off: 0,
-						},
-					},
-				}
-				if err := bcache.PutDirty(
-					file.tailPointer(), file.Branch, fblock); err != nil {
-					return WriteRange{}, nil, err
-				}
 				ptr = fblock.IPtrs[0].BlockPointer
 			}
 
 			// Make a new right block and update the parent's
 			// indirect block list
-			err := fbo.newRightBlockLocked(ctx, lState, file.tailPointer(),
+			err = fbo.newRightBlockLocked(ctx, lState, file.tailPointer(),
 				file.Branch, fblock, startOff+int64(len(block.Contents)), md)
 			if err != nil {
 				return WriteRange{}, nil, err
 			}
+		} else if nCopied < n && off+nCopied < nextBlockOff {
```

Comment:
Hmm, now I'm a little worried about something.  `BlockSplitterSimple.CopyUntilSplit()` currently assumes that the next block starts exactly `maxSize` bytes after this one does.  So I could imagine a scenario where we have these blocks:

```
[0, 512KB], [512KB, 600KB], <hole>, [800KB]
```

Now if someone were to try to write 400KB starting at offset 600KB, `BlockSplitterSimple` would let them, and now the second block would extend past the offset of the final block.  This seems a bit troubling.  We should at least add tests for this case, but we might need to fix it also.  Maybe we can only put holes exactly at block boundaries, or maybe `BlockSplitterSimple` needs to change to take the offset of the next block (if any).